src/components/listbox/listbox.jsx

```diff
@@ -17,18 +18,44 @@ export default function Listbox(props) {
     }
   });
 
+  const refIndexOf = target => props.refs.findIndex(ref => ref.current && ref.current.contains(target));
+  const optionsWasSelected = (toggledRefIndex, isSelected) => {
+    let newSelectedIndexes = selectedIndexes.slice(0, selectedIndexes.length);
+
+    if (isSelected) {
+      newSelectedIndexes.push(toggledRefIndex);
+    } else {
+      const indexOfToggledRef = newSelectedIndexes.indexOf(toggledRefIndex);
+      newSelectedIndexes = newSelectedIndexes.splice(indexOfToggledRef, 1);
+    }
+
+    if (props.selectionLimit && newSelectedIndexes.length > props.selectionLimit) {
+      const deselectIndex = newSelectedIndexes.shift();
+      props.refs[deselectIndex].current.toggleSelected();
+    }
+
+    props.onSelectionChange(props.refs[toggledRefIndex].current.optionData);
```

Comment:
I think it makes more sense to keep the API pretty similar to the select API. Each option has a value. When the option is selected, the select's value is that option's value. 

```suggestion
    props.onSelectionChange(props.refs[toggledRefIndex].current.value);
```

Reply:
I think `onSelectionChange` might be better off if we pass it the whole list of selected values. Usually, when someone is listening to a change event, the value of the target is the whole value -- not just the diff. 

A possibly crazy point, usually when we are listening to an event, we want to pass an event object. This is going to sound wild but check it out: If `Listbox` had an imperative handle then we could `props.onSelectionChange({ target: ref })`. The consumer of `Listbox` would define their listener as;

```
function onListboxSelectionChange(event) {
  if (valid) {
    setValue(event.target.value);
  } else {
    setInvalid(true)
  }
}
```

Actually, perhaps we could simplify the event handler to just `onChange`.

Reply:
```suggestion
  props.onChange({ target: ref });
```

Reply:
To summarize the issue(s):

This does not work when we are consuming DOM events (e.g. `click`), setting state, and then calling `onChange` because the `ref.current` is "stale." 

In order to get this to work, we need to consume DOM events, set state, and on state change call `onChange` -- which ensures the `ref.current` is latest. This is a good pattern to propagate.